[Performance improvement] "Bad tokens ids" optimization

guillaume-be opened the pull request:
Running some benchmarks I noticed that the generation pipeline was varying quite a bit in terms of execution time. Especially the banned token masking seems to be fairly expensive (I ran some experiments where up to 30% of the time for an entire generation process was spent in this step - which seems too high considering its expected simplicity).

This PR accelerates the entire generation pipeline for models using a `bad_words_ids` in their configuration by around 20% on a GPU-enabled node (this includes for example translation using the Marian models). 

The following changes contribute to the performance improvement:
- Single conversion from tensor to list. Previous approach was accessing the GPU buffer for every banned token and every batch element, causing this operation to be slower than the entire forward pass through the model
- Vectorized update of the banned tokens using a masked fill
- Skipping the EOS token for the banned tokens (avoiding a potential duplicate masking)

JetRunner commented on src/transformers/generation_utils.py:
```diff
@@ -89,11 +89,24 @@ def postprocess_next_token_scores(
+                    banned_mask_list.append([idx, token])
+            if len(banned_mask_list) > 0:
+                banned_mask = torch.LongTensor(banned_mask_list)
+                indices = torch.ones(len(banned_mask))
```
`ones_like`?

JetRunner commented on src/transformers/generation_utils.py:
```diff
@@ -89,11 +89,24 @@ def postprocess_next_token_scores(
+            if len(banned_mask_list) > 0:
+                banned_mask = torch.LongTensor(banned_mask_list)
+                indices = torch.ones(len(banned_mask))
+                banned_mask = (
+                    torch.sparse.LongTensor(banned_mask.t(), indices, scores.size())
+                    .to_dense()
+                    .bool()
+                    .to(scores.device)
```
Not 100% understand what's going on here

sshleifer replied:
why not just make a dense tensor to start?

JetRunner replied:
Oh I kinda understand now after reading torch.sparse document

guillaume-be replied:
This is as far as I know the most effective way to create a mask tensor from a list of [idx, pos]. I am not aware of a method to create a dense tensor populated with ones at specific locations, e.g. 
```
[ 0  1  1 ]
[ 0  0  0 ]
[ 1  0  0 ]
```
from: `[[0, 1], [0, 2], [2, 0]]`. As far as I know a dense matrix approach would require a zeroes initialization followed by a for loop to modify the tensor element by element.

JetRunner replied:
Maybe a little comment would help people better understand!

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -89,11 +89,24 @@ def postprocess_next_token_scores(
+                    .bool()
+                    .to(scores.device)
+                )
+                scores.masked_fill_(banned_mask, -float("inf"))
```
love this!

guillaume-be commented:
@sshleifer Thank you very much for the review. I have added unit tests for the modified method that hopefully aligns with what you had in mind. I have re-run the Marian integration tests that run without issue. I somehow have issues running the BART integration tests (even on master) due to an `ImportError` and unable to see if these still run:
```python
from .test_configuration_common import ConfigTester
ImportError: attempted relative import with no known parent package
```

regarding point 2) could you please clarify? The `calc_banned_bad_words_ids` still exists (and is used) in the proposed PR. Would you recommend making a copy of it instead of changing its behaviour? Then the original `calc_banned_bad_words_ids`  would no longer be used anywhere

@JetRunner I have added a comment to clarify the mask tensor generation

Thank you!

sshleifer commented on tests/test_modeling_common.py:
```diff
@@ -971,3 +974,70 @@ def test_top_k_top_p_filtering(self):
+    def test_postprocess_next_token_scores(self):
+
+        config = MarianConfig.from_pretrained("Helsinki-NLP/opus-mt-fr-en")
+        model = MarianMTModel(config=config)
```
you might want to test a smaller model here.

sshleifer replied:
you should definitely test a tiny model since the tiny model doesn't matter.
use `sshleifer/tiny-marian-en-de/`

sshleifer commented on tests/test_modeling_common.py:
```diff
@@ -971,3 +974,70 @@ def test_top_k_top_p_filtering(self):
+                32,
+                5,
+            )
+        except timeout_decorator.timeout_decorator.TimeoutError:
```
This is meant to test that the code stays fast, so you should let it raise. Ideally it is a test that would have failed using the old logic.

sshleifer commented:
I misread your code, sorry.
My point 2 should be that it feels like the new masking logic could be put into a helper method like
```python
def set_scores_to_inf_for_banned_tokens(self, scores, bad_words_ids) -> None:

```

What is the speedup?

guillaume-be commented:
@sshleifer This makes sense, just pushed a few more changes:
- Moved the masking to a utility function
- Updated the unit test to let it fail if it hits timeout. As this is configuration dependent, the limit was increased to 10 if the CI compute power available fluctuates. In general I am not sure if unit tests are the best way to perform performance regression tests
- I have created a gist to share the performance difference between the current and the proposed approach: https://gist.github.com/guillaume-be/e335b099005e9bf38448d0e2eb02f74f . On this simple example with a GPU on Colab, the proposed approach is twice as fast. This actually has a significant impact on the entire generation process, but I did not manage to create a good example on Colab (the resources fluctuate too much from notebook to notebook, and not aware of a way to change a library version within a same notebook). Running locally with a consumer-grade Turing GPU (2070), I observe a time reduction of around 20% for the end-to-end generation process.

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -89,11 +89,11 @@ def postprocess_next_token_scores(
-            for i, banned_tokens in enumerate(banned_tokens):
-                scores[i, banned_tokens] = -float("inf")
+            banned_tokens = calc_banned_bad_words_ids(input_ids.tolist(), bad_words_ids)
+            set_scores_to_inf_for_banned_tokens(scores, banned_tokens)
```
comment that this modifies the `scores` tensor

sshleifer commented on src/transformers/generation_utils.py:
```diff
@@ -904,6 +904,25 @@ def _tokens_match(prev_tokens, tokens):
+    for idx, batch_banned_tokens in enumerate(banned_tokens):
+        for token in batch_banned_tokens:
+            banned_mask_list.append([idx, token])
+    if len(banned_mask_list) > 0:
```
(nit)
```python
if not banned_mask_list:
    return
```
will save some indentation

sshleifer commented on tests/test_modeling_common.py:
```diff
@@ -971,3 +974,67 @@ def test_top_k_top_p_filtering(self):
         self.assertTrue(torch.all(torch.eq(non_inf_expected_idx, non_inf_idx)))
+
+    @require_torch
+    def test_postprocess_next_token_scores(self):
```
I'm worried that this is going to run for every model. Maybe it should be in it's own `test_generation_utils.py` file?

sshleifer commented on tests/test_modeling_common.py:
```diff
@@ -971,3 +974,67 @@ def test_top_k_top_p_filtering(self):
+        masked_scores = [
+            [(0, 299), (1, 299), (2, 299), (3, 299), (4, 299), (5, 299), (6, 299), (7, 299)],
+            [(1, 24), (0, 54), (1, 54), (2, 54), (3, 54), (4, 54), (5, 54), (6, 54), (7, 54)],
+            [(0, 0), (1, 0), (2, 0), (3, 0), (4, 0), (5, 0), (6, 0), (7, 0)],
```
the 0th token is masked by default? is that a special token? if so lets get it from `config.eos_token_id`

this could be more readable if it were [(i,0) for i in range(batch_size)] (same for 988 and everything but first entry of 989.

sshleifer commented on tests/test_modeling_common.py:
```diff
@@ -971,3 +974,67 @@ def test_top_k_top_p_filtering(self):
+    @timeout_decorator.timeout(10)
+    def test_postprocess_next_token_scores_large_bad_words_list(self):
+
+        config = MarianConfig.from_pretrained("Helsinki-NLP/opus-mt-fr-en")
```
you can use 
```python
@cached_property
def model(self): return MarianModel(MarianConfig.from_pretrained("Helsinki-NLP/opus-mt-fr-en"))
```
to only initialize once. Doesn't matter as much if you just use a small model, but a nice trick

guillaume-be commented:
@sshleifer Thank you again for the thorough review! Tried to address the latest comments - I believe it cleans it up quite a bit thank you for the suggestions

sshleifer commented on src/transformers/data/test_generation_utils.py:
```diff
@@ -0,0 +1,95 @@
+        config = MarianConfig.from_pretrained("sshleifer/tiny-marian-en-de")
+        return config, MarianMTModel(config)
+
+    @require_torch
```
if the class requires torch, I don't think each test needs to.

sshleifer commented on src/transformers/data/test_generation_utils.py:
```diff
@@ -0,0 +1,95 @@
+        config, model = self.config_and_model
+        # Initialize an input id tensor with batch size 8 and sequence length 12
+        input_ids = torch.arange(0, 96, 1).view((8, 12))
+
```
```suggestion
eos = config.eos_token_id
bs = input_ids.shape[0]

scores_that_should_be_inf = [[(i, 299) for i in range(bs)], 
+[(i, 54) for i in range(bs)] +[(1,24)], 
[(i, eos) for i in range(bs)], 
]


```
why only (1,24)?

sshleifer replied:
btw that suggestion wont work perfectly

guillaume-be replied:
You are right - removed the unnecessary `@require_torch` and simplified the cached properties into 2 independent properties. For the second test, only the (1, 24) is `-inf` because the banned_token is a sequence: [23, 24]. Since the `input_ids` contain the token id 23 only on batch 1 (by construction), the token is only banned for this batch.

sshleifer commented:
This is ready to be merged @LysandreJik !
